### PR TITLE
add support for Organizational Units

### PR DESCRIPTION
**Pull request by afeld**
Part of #571. Builds on #4229 - [diff](https://github.com/afeld/terraform-provider-aws/compare/root-ou...ou).

This pull request will add support for [Organizational Units](https://docs.aws.amazon.com/organizations/latest/userguide/orgs_manage_ous.html). ~~Still needs work - mostly wanted to submit to let folks know I'm working on it.~~

* [x] Get basic acceptance tests passing
* [x] Add update logic
* [x] Add the ability to specify a parent OU
* [x] Rebase

### CODE REVIEW

**Comment by afeld** [comment]
Ready for review! I'll rebase after/if #571 is merged to reduce the size of the pull request. Thanks!

**Comment by afeld** [comment]
```
$ make testacc TEST=./aws TESTARGS="-run TestAccAWSOrganizations/Unit"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run TestAccAWSOrganizations/Unit -timeout 120m
=== RUN   TestAccAWSOrganizations
=== RUN   TestAccAWSOrganizations/Unit
=== RUN   TestAccAWSOrganizations/Unit/basic
=== RUN   TestAccAWSOrganizations/Unit/importBasic
=== RUN   TestAccAWSOrganizations/Unit/update
--- PASS: TestAccAWSOrganizations (49.83s)
    --- PASS: TestAccAWSOrganizations/Unit (49.83s)
        --- PASS: TestAccAWSOrganizations/Unit/basic (15.74s)
        --- PASS: TestAccAWSOrganizations/Unit/importBasic (13.32s)
        --- PASS: TestAccAWSOrganizations/Unit/update (20.76s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	64.820s
```

File: aws/resource_aws_organizations_unit.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringLenBetween(1, 128),
+			},
+			"parent_id": {
+				ForceNew:     true,
```
Comment (afeld) [inline aws/resource_aws_organizations_unit.go]: Can add support for moving Organizational Units non-destructively later.

**Comment by n3ph** [comment]
We need this to attach SCPs to groups.. Would be nice to see this merged

**Comment by bentterp** [comment]
Any chances of getting this merged soon? OUs are a central part of the whole Organizations concept.

**Comment by glerma** [comment]
Bump?

**Comment by zekefast** [comment]
@afeld Are conflicts prevent this PR from merging? Would you be able to rebase? I can help if you have no time.

**Comment by pstrav** [comment]
@afeld Any news on this one? Would it be of great help :)

**Comment by zekefast** [comment]
@pstrav  It is dead as a hell. I tried to reach people in Gitter channels, but ignorance everywhere and number of PRs to the repo are growing. When I was started to monitor issues related to `aws_organizations` and SCP there was `205` PRs in the repo. Now it is `233`. And I'm not sure that maintainers will prioritize `organizations` service at all. :(

**Comment by jtreutel** [comment]
I really hope this gets merged soon, lack of OU support is the only reason we are not using Terraform to manage our AWS Organization.  It's a core part of AWS Org's functionality.

**Comment by ndobbs** [comment]
This is very sorely needed. We can't move forward with our terraform projects until Organizational Units are supported.

**Comment by joshuatalb** [comment]
Any chance of getting this merged in? This is a huge blocker for us at the moment. (cc: @bflad @afeld)

**Comment by Comradin** [comment]
Hello, any chance to get this one merged anytime soon?

Is there anything we can do to provide help fixing this issue?

**Comment by bflad** [comment]
Can this PR please have the resource and data source split? I will prioritize getting the resource reviewed and merged, but the various pull requests for Organizations data sources will all need to be reviewed separately since they each implement different workflows. Thanks!

**Comment by ndobbs** [comment]
@bflad Is this OK to to be reviewed now? Thanks a ton.

**Comment by JuanOI** [comment]
Really looking forward to this being released!

**Comment by bflad** [comment]
To provide an update, the maintainers will be circling back to looking through the various open Organizations data source pull requests to review them all at once after we determine which make the most sense for expected use cases.

> @bflad You saw #4229? That's the data source on its own - I built this pull request on that one so that I could use the data source in the tests. I was thinking I would rebase this PR once the other is merged. That work?

Yes, I saw that pull request. 😄 My point is more centered around that data sources should not be _required_ to provision infrastructure. That type of configuration within Terraform is discouraged where possible and the acceptance testing (analogous to real world configurations) should be able to stand up an Organization with Organizational Units from the ground up with resources only.

In this case, it seems like we should be able to add the Root ID(s) in the `aws_organizations_organization` resource to remove the data source requirement. Once this is done, the resource testing here references only resources and the data source testing references this resource to match the testing in the rest of the provider.

Hope this makes sense.

**Comment by bryanlalexander** [comment]
@bflad Is your contention that the PR from @afeld requires the data source in order to call the resource, and the solution is to return the values from a call to `organizations:ListRoots` in the Read and Create from `aws_organizations_organizations`?

**Comment by bflad** [comment]
@bryanlalexander that is spot on. 💯 If there isn't any movement on that implementation in the next few days I can likely get that submitted. Apologies this process has been dragging out for so long.

**Comment by Midacts** [comment]
What are the next steps to get this added?    : o

**Comment by bflad** [comment]
The `aws_organizations_organization` resource now has a `roots` attribute, which we can leverage in our acceptance testing and in the real world via `"${aws_organizations_organization.XXX.roots.0.id}"` 👍 

@afeld please let us know if you can update this pull request to just the resource and we can provide quick turnaround on subsequent reviews. Thanks so much.

**Comment by afeld** [comment]
Sorry for the slow response, folks. I will try and get to it later this weekend, but can't promise. Also, haven't been working with Go or AWS Organizations for a while, so I'll be a bit rusty. If anyone else wants to pick it up and run with it in the meantime, you are more than welcome. Just reach out and I'll make you a committer on my fork.

**Comment by bryanlalexander** [comment]
@bflad we should be good to go now.

**Comment by bflad** [comment]
@afeld thank you so much for the heads up. We are all really appreciative of your work that went into this. Apologies again this was held up for so long.

@bryanlalexander 💯 🥇 thanks for those updates! Taking a deeper look now, but overall looks to be in pretty good shape.

**Comment by bflad** [comment]
This has been released in [version 2.10.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2100-may-10-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**